### PR TITLE
usb: make usb descriptor power options configurable

### PR DESCRIPTION
**Pull request by solsbarry**
Add two new kconfig options USB_SELF_POWERED and USB_MAX_POWER.
These can be set by the user to change the USB configuration descriptor.
USB_MAX_POWER can be set to any value between 0 and 250, but practically
should be 50 or 250. These values are half the ammount of mA that the
device will tell the host that it needs.
USB_SELF_POWERED sets the 7th bit in bmAttributes of the USB config 
descriptor. Should be set to y is the device has its own power source
 other than USB.

Proposed fix for https://github.com/zephyrproject-rtos/zephyr/issues/19545

Signed-off-by: Barry Solomon barry.solomon@dexcom.com

### CODE REVIEW

File: subsys/usb/Kconfig [hunk 1]
```diff
@@ -88,6 +88,17 @@ config USB_DEVICE_BOS
...
+config USB_SELF_POWERED
+    bool "Set the power parameters for the usb configuration descriptor to indicate self powered"
+    default y
+    
+config USB_MAX_POWER
+    int "Set the max power in the usb configuration descriptor"
+    default 50
+    range 0 255
+    help
+      The result is 2mA times the vlaue provided.
```
Comment (jfischer-phytec-iot) [inline subsys/usb/Kconfig]: value

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Please also squash to one commit.

**Comment by solsbarry** [comment]
I made all the changes John recommended and squashed to a single commit.

File: subsys/usb/usb_descriptor.c [hunk 2]
```diff
@@ -88,7 +88,7 @@ USBD_DEVICE_DESCR_DEFINE(primary) struct common_descriptor common_desc = {
 		.bConfigurationValue = 1,
 		.iConfiguration = 0,
 		.bmAttributes = USB_CONFIGURATION_ATTRIBUTES,
-		.bMaxPower = MAX_LOW_POWER,
+		.bMaxPower = CONFIG_USB_MAX_POWER,
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_descriptor.c]: please do the same for `bMaxPower = MAX_LOW_POWER` in subsys/usb/class/usb_dfu.c
please remove `#define MAX_LOW_POWER                  0x32` from include/usb/usb_common.h
```diff
diff --git a/include/usb/usb_common.h b/include/usb/usb_common.h
index 632524a464..5fa617e098 100644
--- a/include/usb/usb_common.h
+++ b/include/usb/usb_common.h
@@ -86,11 +86,6 @@
 #define BCDDEVICE_RELNUM               (BCD(KERNEL_VERSION_MAJOR) << 8 | \
                                        BCD(KERNEL_VERSION_MINOR))
 
-/* 100mA max power, per 2mA units */
-/* USB 1.1 spec indicates 100mA(max) per unit load, up to 5 loads */
-#define MAX_LOW_POWER                  0x32
-#define MAX_HIGH_POWER                 0xFA
-
 /* Highest value of Frame Number in SOF packets. */
 #define USB_SOF_MAX                    2047
 
diff --git a/subsys/usb/class/usb_dfu.c b/subsys/usb/class/usb_dfu.c
index bc3169f0e0..750c5c936f 100644
--- a/subsys/usb/class/usb_dfu.c
+++ b/subsys/usb/class/usb_dfu.c
@@ -156,7 +156,7 @@ struct dev_dfu_mode_descriptor dfu_mode_desc = {
                .bConfigurationValue = 1,
                .iConfiguration = 0,
                .bmAttributes = USB_CONFIGURATION_ATTRIBUTES,
-               .bMaxPower = MAX_LOW_POWER,
+               .bMaxPower = CONFIG_USB_MAX_POWER,
        },
        .sec_dfu_cfg = {
                /* Interface descriptor */

```

File: include/usb/usb_common.h [hunk 3]
```diff
@@ -101,9 +101,11 @@
...
  */
 #define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP  0x20
-#define USB_CONFIGURATION_ATTRIBUTES                0xC0       \
-	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,        \
-	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0)))
+#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED   0xC0
+#define USB_CONFIGURATION_ATTRIBUTES ((COND_CODE_1(USB_SELF_POWERED, \
+	   (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
+	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,         \
+	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))
```
Comment (jfischer-phytec-iot) [inline include/usb/usb_common.h]: (0) does not make any sense

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: include/usb/usb_common.h (hunk 3 above)
Comment (solsbarry) [reply include/usb/usb_common.h]: That was in the original code, so I left it and emulated it, but I agree it makes no sense.

File: include/usb/usb_common.h [hunk 4]
```diff
@@ -101,9 +101,11 @@
  * D4...0:Reserved -> 0
  */
 #define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP  0x20
```
Comment (emob-nordic) [inline include/usb/usb_common.h]: In fact USB_CONFIGURATION_ATTRIBUTES is bitmask so i would personally go for:

```diff
diff --git a/include/usb/usb_common.h b/include/usb/usb_common.h
index 632524a464..d2220a9692 100644
--- a/include/usb/usb_common.h
+++ b/include/usb/usb_common.h
@@ -100,9 +100,10 @@
  * D5:Remote Wakeup -> 0,
  * D4...0:Reserved -> 0
  */
-#define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP  0x20
-#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED   0xC0
-#define USB_CONFIGURATION_ATTRIBUTES ((COND_CODE_1(USB_SELF_POWERED, \
+#define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP	BIT(5)
+#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED	BIT(6)
+#define USB_CONFIGURATION_ATTRIBUTES BIT(7) \
+	| ((COND_CODE_1(USB_SELF_POWERED, \
 	   (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
 	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,         \
 	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))

```
As @jfischer-phytec-iot  mentioned BIT(7) must be set (historical reasons - could be found in usb20 spec)

**Review by eobalski — CHANGES_REQUESTED** [review]
"Should be set to y is the device" -> "Should be set to y if the device" in the commit message

File: include/usb/usb_common.h (hunk 4 above)
Comment (emob-nordic) [reply include/usb/usb_common.h]: (+) s/USB_SELF_POWERED/CONFIG_USB_SELF_POWERED as @jfischer-phytec-iot mentioned. I forgot about that.

File: subsys/usb/usb_descriptor.c (hunk 2 above)
Comment (solsbarry) [reply subsys/usb/usb_descriptor.c]: will do.

File: include/usb/usb_common.h [hunk 5]
```diff
@@ -101,9 +101,11 @@
...
-#define USB_CONFIGURATION_ATTRIBUTES                0xC0       \
-	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,        \
-	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0)))
+#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED   0xC0
+#define USB_CONFIGURATION_ATTRIBUTES ((COND_CODE_1(USB_SELF_POWERED, \
+	   (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
```
[the remark this answers is not in the document]
Comment (solsbarry) [reply include/usb/usb_common.h]: will do

File: include/usb/usb_common.h (hunk 4 above)
Comment (solsbarry) [reply include/usb/usb_common.h]: this makes sense

Comment (solsbarry) [reply include/usb/usb_common.h]: And on the parentheses around the zeo in the 3rd argument of COND_CODE_1 that seems pretty standard. I see it in other places in the repo. Sometimes the argument is even empty parentheses.

**Review by eobalski — APPROVED** [review]
lgmt now.

File: include/usb/usb_common.h [hunk 6]
```diff
@@ -100,10 +95,13 @@
  * D5:Remote Wakeup -> 0,
  * D4...0:Reserved -> 0
  */
-#define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP  0x20
-#define USB_CONFIGURATION_ATTRIBUTES                0xC0       \
-	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,        \
-	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0)))
+#define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP	BIT(5)
```
Comment (emob-nordic) [inline include/usb/usb_common.h]: I will consider one more thing: 
```diff
diff --git a/include/usb/usb_common.h b/include/usb/usb_common.h
index c3c0058e2e..7aa98c3d8e 100644
--- a/include/usb/usb_common.h
+++ b/include/usb/usb_common.h
@@ -95,13 +95,9 @@
  * D5:Remote Wakeup -> 0,
  * D4...0:Reserved -> 0
  */
-#define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP	BIT(5)
-#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED	BIT(6)
 #define USB_CONFIGURATION_ATTRIBUTES BIT(7) \
-	| ((COND_CODE_1(CONFIG_USB_SELF_POWERED, \
-	   (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
-	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,         \
-	   (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))
+	| (COND_CODE_1(CONFIG_USB_SELF_POWERED, (BIT(6)), (0)))   \
+	| (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP, (BIT(5)), (0)))
 
 /* Classes */
 #define COMMUNICATION_DEVICE_CLASS	0x02

```
As those macros are not used anywhere else.

Comment (jfischer-phytec-iot) [reply include/usb/usb_common.h]: No please do not remove this defines, it makes code more readable also could be used in the stack or application (it is public API ).

File: include/usb/usb_common.h [hunk 7]
```diff
@@ -100,10 +95,13 @@
...
+#define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED	BIT(6)
+#define USB_CONFIGURATION_ATTRIBUTES BIT(7) \
```
Comment (jfischer-phytec-iot) [inline include/usb/usb_common.h]: I would indent this line with tabs to be aligned with the lines above.
```diff
diff --git a/include/usb/usb_common.h b/include/usb/usb_common.h
index c3c0058e2e..1dba1d163b 100644
--- a/include/usb/usb_common.h
+++ b/include/usb/usb_common.h
@@ -97,11 +97,11 @@
  */
 #define USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP     BIT(5)
 #define USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED      BIT(6)
-#define USB_CONFIGURATION_ATTRIBUTES BIT(7) \
-       | ((COND_CODE_1(CONFIG_USB_SELF_POWERED, \
-          (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
-       | (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,         \
-          (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))
+#define USB_CONFIGURATION_ATTRIBUTES                   BIT(7) \
+               | ((COND_CODE_1(CONFIG_USB_SELF_POWERED, \
+                  (USB_CONFIGURATION_ATTRIBUTES_SELF_POWERED), (0)))   \
+               | (COND_CODE_1(CONFIG_USB_DEVICE_REMOTE_WAKEUP,         \
+                  (USB_CONFIGURATION_ATTRIBUTES_REMOTE_WAKEUP), (0))))

```

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**